Allow macro config to come from plugin options in babel config.

conartist6 opened the pull request:
Fixes #112 

I'm missing test coverage, but I'm also not familiar with the way this test environment is set up. Some tips on how to test this well would be appreciated.

conartist6 commented on other/docs/author.md:
```diff
@@ -4,9 +4,9 @@
-* [Writing custom Babel and ESLint plugins with ASTs](https://youtu.be/VBscbcm2Mok?list=PLV5CVI1eNcJgNqzNwcs4UKrlJdhfDjshf): A 53 minute talk by [@kentcdodds](https://twitter.com/kentcdodds)
-* [babel-handbook](https://github.com/thejameskyle/babel-handbook): A guided handbook on how to use Babel and how to create plugins for Babel by [@thejameskyle](https://twitter.com/thejameskyle)
-* [Code Transformation and Linting](https://kentcdodds.com/workshops/#code-transformation-and-linting): A workshop (recording available on Frontend Masters) with exercises of making custom Babel and ESLint plugins
+- [Writing custom Babel and ESLint plugins with ASTs](https://youtu.be/VBscbcm2Mok?list=PLV5CVI1eNcJgNqzNwcs4UKrlJdhfDjshf): A 53 minute talk by [@kentcdodds](https://twitter.com/kentcdodds)
```
I don't know why it did this. Maybe you hadn't run the newest version of your `kcd-scripts` formatter on this file?

conartist6 commented:
Never mind I figured out how to write the tests. Should be ready to go.

conartist6 commented:
Whoops, forgot to git add a folder. Should really be ready to go.

kentcdodds commented on src/index.js:
```diff
@@ -240,32 +255,55 @@ function getConfig(macro, filename, source) {
       if (loaded) {
-        return loaded.config[macro.options.configName]
+        configOptions = loaded.config[configName]
+        configPath = loaded.filepath
```
Could we break up things a bit into a `getMacroConfigFromConfigFile` and a `getMacroConfigFromBabelConfig`? I fee like having these mixed up makes it more complicated than it needs to be.

conartist6 replied:
Sure

conartist6 commented on src/index.js:
```diff
@@ -228,9 +230,22 @@ function applyMacros({
+    }
+
     try {
       // lazy-loading it here to avoid perf issues of loading it up front.
```
So I appreciate the comment, but maybe phobia was the right word for what influenced this decision. First, if you must synchronously load `cosmiconfig` no matter what I don't see how you can gain anything useful by doing it later, unless of course it turns out that babel-plugin-macros is loaded but no macros are used (which hardly seems like the case to optimize for).

Second, in optimizing for that case, you've actually hurt performance for the common case significantly. If you read the cosmiconfig docs you'll see that its caches are internal to the `explorer` object. You create a new explorer every time you process a file, so you are also doing a full config search every time you process a file, which is far more costly in the common case than what you were trying to avoid.

It's easy to refactor the code, but it seems like it'll be a bit tricker to fix the tests.

conartist6 replied:
Actually I suppose we can have the best of both worlds if we lazy load and cache the explorer

conartist6 replied:
This is done now.

conartist6 commented:
I've added some more or less unrelated changes, but I've split them out into separate commits.

When trying to debug the tests I discovered that most tests assertion output was silently discarded.
I also found that failing objectContaining assertions where the actual object contains babel were printing output so long that they were obscuring useful output by pushing it completely out of my 1000 line terminal scrollback.

kentcdodds commented on src/index.js:
```diff
@@ -228,44 +243,76 @@ function applyMacros({
+}
+
+function getConfigFromOptions(configName, options) {
+  if (Object.prototype.hasOwnProperty.call(options, configName)) {
```
Any reason this can't just be `options.hasOwnProperty(configName)`?

conartist6 replied:
Up to you. It's not a wholly safe assumption -- it is violated by `options = Object.create(null)`. If that's acceptable to you I'll change it.

kentcdodds replied:
I've literally never seen anyone create an object like that in real code :)

conartist6 replied:
I got burned by it writing the `entries` method for iter-tools. Usually the only such objects you'll encounter are internal data structures though, never configurations. I'll change it. Any thoughts on the rest of the changes?

kentcdodds commented on other/docs/author.md:
````diff
@@ -164,43 +164,53 @@ This is a string used as import declaration's source - i.e. `'./my.macro'`.
-
-```javascript
-// babel-plugin-macros.config.js
+// .babel.config.js
````
I'd prefer `babel.config.js` and `babel-plugin-macros.config.js` rather than `.babel.config.js` and `.babel-plugin-macros.config.js`

kentcdodds commented on other/docs/author.md:
````diff
@@ -164,43 +164,53 @@ This is a string used as import declaration's source - i.e. `'./my.macro'`.
 ```
 
-And the `config` object you would receive would be: `{someConfig: {}}`.
+Note that in the above example if both files were sepcified the finale locale value would
````
`finale` should be `final`

kentcdodds commented on other/docs/author.md:
```diff
@@ -164,43 +164,53 @@ This is a string used as import declaration's source - i.e. `'./my.macro'`.
 
-And the `config` object you would receive would be: `{someConfig: {}}`.
+Note that in the above example if both files were sepcified the finale locale value would
+be `en_GB`, since that is the value in the plugin options.
```
Are we sure we want the plugin option to override the config version. It feels more intuitive to me that it would be the other way around.

kentcdodds commented:
@all-contributors please add @conartist6 for code, tests, and docs

kentcdodds commented:
:tada: This PR is included in version 2.6.0 :tada:

The release is available on:
- `npm package (@latest dist-tag)`
- [GitHub release](https://github.com/kentcdodds/babel-plugin-macros/releases/tag/v2.6.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

conartist6 commented:
Awesome! Thanks! By the way, I realized that something I said was wrong. It definitely does make sense for macros to optimize for the case where it isn't being used, because one of its goals is to allow people who don't have control over their transpilation environment access to the power of transformations. That means it is a reasonable expectation that it will be added to quite a few projects that won't use it at all.

InvictusMB commented on src/index.js:
```diff
@@ -88,6 +108,7 @@ function macrosPlugin(
               babel,
               interopRequire,
               resolvePath,
+              options,
```
@conartist6 @kentcdodds 
Shouldn't the `options` also be passed in another `applyMacros` invocation?

https://github.com/kentcdodds/babel-plugin-macros/blob/d11217e724ca00125ead59f82a564776d61b56bc/src/index.js#L146-L154

kentcdodds replied:
Could you open a new issue and/or a pull request with a failing test? Thanks!

conartist6 replied:
@InvictusMB You are right. It would affect macros meant to be configured via plugin options which are imported with `const { sym } = require('path/to/macro')`. Nice catch.

conartist6 replied:
I have created #120 with the repro and fix for this issue.
